### DISCUSSION TITLE
Mesmer (version 0.8.0) - How to collect Default-JVM-Metrics from ZIO 2.0.0 Apps

### DISCUSSION THREAD

**Discussion by ghost** [ScalaConsultants/mesmer#458] [Show and tell]
This Mesmer Discussion is related to this "Mesmer Research Task (https://github.com/ScalaConsultants/mesmer/issues/424)".

I will be answering the following questions from https://github.com/ScalaConsultants/mesmer/issues/424:

1. "Figure out how to turn on Jvm metrics collection."
2. "If it works with few configuration needed, explain how to do it in mesmer documentation so that users can see those thanks to Mesmer as well"

_**1) Figure out how to turn on Jvm metrics collection**_

- In the User's ZIO 2.0 App, make sure that this ZIO App has the following two lines added in the code: https://github.com/zio/zio-metrics-connectors/blob/zio/series2.x/core/jvm/src/test/scala/zio/metrics/connectors/SampleApp.scala#L69-L70

- Both `Runtime.enableRuntimeMetrics` are `DefaultJvmMetrics.live.unit` needed, in order to enable the ZIO internal metrics and the default JVM metricsConfig.

**_2. "If it works with few configuration needed, explain how to do it in mesmer documentation so that users can see those thanks to Mesmer as well"_**

- As of now, Mesmer is **not able** to collect ZIO DefaultJvmMetrics.
- However, ZIO 2.0 Devs are currently working on a feature regarding "collecting JVM Metrics inside a ZIO Application": https://zio.dev/reference/metrics/jvm/#collecting-metrics
- I will keep an eye for updates to this ZIO JVM Metrics documentation (https://zio.dev/reference/metrics/jvm/), and will update this discussion as needed.

**Comment by ghost** [comment]
CC @lgajowy and @worekleszczy

**Comment by lgajowy** [comment]
> As of now, Mesmer is not able to collect ZIO DefaultJvmMetrics.

@ronaldSC-2022 we recently merged zio counters and gauges collection. If you run mesmer from main branch with the forkenstein app (that has the default jvm metrics enabled) doesn't it collect the counters and gauges?

**Reply by ghost** [reply]
@lgajowy, you are correct:

If we run Mesmer from the main branch with the forkenstein app (that has this `DefaultJvmMetrics.live.unit` commit: https://github.com/lgajowy/forkenstein/pull/2/commits/ce557492b468aeb26ad5c3e326e30432177f1941)... then Mesmer will collect the counters and gauges.

I think I mis-spoke / I wrote my discussion incorrectly:

-  "As of now, Mesmer is not able to collect ZIO DefaultJvmMetrics" should be re-written as "Mesmer is not able to collect ZIO DefaultJvmMetrics **_if the User's ZIO App does not have the default jvm metrics enabled"_**

Everything else, I agree with you. 😄

PS, if needed, let's discuss more in our Mesmer meeting today, Lukasz.

**Comment by lgajowy** [comment]
I confirm that we get them forwarded to otel: 
<img width="916" alt="Screenshot 2022-08-10 at 13 07 11" src="https://user-images.githubusercontent.com/1932045/183886798-f0a7d4c4-3b08-4048-b190-b21d18b2ce61.png">
<img width="1683" alt="Screenshot 2022-08-10 at 13 07 34" src="https://user-images.githubusercontent.com/1932045/183886811-bd5e7a0a-d357-4934-ae61-95f53f50aaf1.png">

Great stuff :)
